File: java/affine-cipher/src/main/java/AffineCipher.java
import java.math.BigInteger;

public class AffineCipher {
  static final int M = 26;

  String encode(String plain, int a, int b) {
    checkValidity(a);

    StringBuilder result = new StringBuilder();
    for (char c : plain.toLowerCase().toCharArray()) {
      if (Character.isLetterOrDigit(c)) {
        char encoded = Character.isLetter(c) ? (char) ((a * (c - 'a') + b) % M + 'a') : c;

        if (result.length() % 6 == 5) {
          result.append(' ');
        }
        result.append(encoded);
      }
    }

    return result.toString();
  }

  String decode(String cipher, int a, int b) {
    checkValidity(a);

    StringBuilder result = new StringBuilder();
    for (char c : cipher.toCharArray()) {
      if (Character.isLetterOrDigit(c)) {
        char decoded =
            Character.isLetter(c)
                ? (char) (multiplyMod(modInverse(a), Math.floorMod((c - 'a') - b, M)) + 'a')
                : c;

        result.append(decoded);
      }
    }

    return result.toString();
  }

  int modInverse(int x) {
    return BigInteger.valueOf(x).modInverse(BigInteger.valueOf(M)).intValue();
  }

  int multiplyMod(int x, int y) {
    return x * y % M;
  }

  void checkValidity(int a) {
    if (gcd(a, M) != 1) {
      throw new IllegalArgumentException("Error: keyA and alphabet size must be coprime.");
    }
  }

  int gcd(int x, int y) {
    return (y == 0) ? x : gcd(y, x % y);
  }
}

File: python/affine-cipher/affine_cipher.py
import math

M = 26


def encode(plain_text, a, b):
    check_validity(a)

    result = ''
    for ch in plain_text.lower():
        if ch.isalnum():
            if ch.isalpha():
                encoded = chr((a * (ord(ch) - ord('a')) + b) % M + ord('a'))
            else:
                encoded = ch

            if len(result) % 6 == 5:
                result += ' '
            result += encoded

    return result


def decode(ciphered_text, a, b):
    check_validity(a)

    result = ''
    for ch in ciphered_text:
        if ch.isalnum():
            if ch.isalpha():
                encoded = chr(mod_inv(a) * (ord(ch) - ord('a') - b % M + M) % M
                              + ord('a'))
            else:
                encoded = ch

            result += encoded

    return result


def mod_inv(a):
    result = 1
    while result * a % M != 1:
        result += 1

    return result


def check_validity(a):
    if math.gcd(a, M) != 1:
        raise ValueError('A and alphabet size (26) must be coprime!')
